services/mac_manager/src/main/java/com/futurewei/alcor/macmanager/service/implement/MacServiceImpl.java

```diff
@@ -62,236 +65,337 @@
         if (macAddress.equals(macState.getMacAddress()) == false)
-            throw (new ParameterUnexpectedValueException(MacUtil.MAC_EXCEPTION_PARAMETER_INVALID));
-        if (macStateRepository.findItem(macAddress) != null) {
-            macStateRepository.addItem(macState);
-        } else {
-            throw (new ResourceNotFoundException(MacUtil.MAC_EXCEPTION_MAC_NOT_EXISTING));
+            throw (new ParameterUnexpectedValueException(MacManagerConstant.MAC_EXCEPTION_PARAMETER_INVALID));
+        try {
+            if (macStateRepository.findItem(macAddress) != null) {
+                macStateRepository.addItem(macState);
+            } else {
+                ResourceNotFoundException e = new ResourceNotFoundException(MacManagerConstant.MAC_EXCEPTION_MAC_NOT_EXISTING);
+                logger.error("MacService updateMacState() exception:", e);
+                throw (e);
+            }
+        } catch (CacheException e) {
+            throw new MacRepositoryTransactionErrorException(MacManagerConstant.MAC_EXCEPTION_REPOSITORY_EXCEPTION);
         }
         return macState;
     }
 
-    public String releaseMacState(String macAddress) throws Exception {
+    /**
+     * release MAC address from an allocation and put it back to MAC address pool
+     *
+     * @param macAddress MAC address
+     * @return MAC address
+     * @throws ParameterNullOrEmptyException          parameter macAddress is null or empty
+     * @throws MacRepositoryTransactionErrorException error during repository transaction
+     * @throws ResourceNotFoundException              there is not mac state to release macAddress
+     */
+    @Override
+    public String releaseMacState(String macAddress) throws ParameterNullOrEmptyException, MacRepositoryTransactionErrorException, ResourceNotFoundException {
+        String strMacAddress = null;
         if (macAddress == null)
-            throw (new ParameterNullOrEmptyException(MacUtil.MAC_EXCEPTION_PARAMETER_NULL_EMPTY));
-        MacState macState = macStateRepository.findItem(macAddress);
-        if (macState == null) {
-            throw (new ResourceNotFoundException(MacUtil.MAC_EXCEPTION_MAC_NOT_EXISTING));
-        } else {
-            try {
-                macStateRepository.deleteItem(macAddress);
-                macPoolRepository.addItem(new MacAddress(macAddress));
-            } catch (Exception e) {
-                throw e;
+            throw (new ParameterNullOrEmptyException(MacManagerConstant.MAC_EXCEPTION_PARAMETER_NULL_EMPTY));
+        try {
+            MacState macState = macStateRepository.findItem(macAddress);
+            if (macState == null) {
+                throw (new ResourceNotFoundException(MacManagerConstant.MAC_EXCEPTION_MAC_NOT_EXISTING));
+            } else {
+                try {
+                    macStateRepository.deleteItem(macAddress);
+                    MacAddress mac = new MacAddress(macAddress);
```

Comment:
These three lines could be grouped as a method like "inactivateMacAddressBit" or so.

Reply:
These three lines are grouped in a method :  private void inactivateMacAddressBit(String macAddress) throws MacRepositoryTransactionErrorException